packages/flutter/lib/src/widgets/pop_scope.dart

```diff
@@ -0,0 +1,142 @@
+/// back gestures when using nested [Navigator]s.
+///
+/// ** See code in examples/api/lib/widgets/pop_scope/pop_scope.0.dart **
+/// {@end-tool}
+///
+/// See also:
+///
+///  * [NavigatorPopHandler], which is a less verbose way to handle system back
+///    gestures in the case of nested [Navigator]s.
+///  * [Form.popEnabled] and [Form.onPopped], which can be used to handle system
+///    back gestures in the case of a form with unsaved data.
+///  * [ModalRoute.registerPopInterface] and [ModalRoute.unregisterPopInterface],
+///    which this widget uses to integrate with Flutter's navigation system.
+class PopScope extends StatefulWidget implements PopInterface {
+  /// Creates a widget that registers a callback to veto attempts by the user to
+  /// dismiss the enclosing [ModalRoute].
+  ///
+  /// The [child] argument must not be null.
+  const PopScope({
+    super.key,
+    required this.child,
+    this.popEnabled = true,
+    this.onPopped,
+  });
+
+  /// The widget below this widget in the tree.
+  ///
+  /// {@macro flutter.widgets.ProxyWidget.child}
+  final Widget child;
+
+  /// Called after a route pop was handled.
+  ///
+  /// It's not possible to prevent the pop from happening at the time that this
+  /// method is called; the pop has already happened. Use [popEnabled] to
+  /// disable pops in advance.
+  ///
+  /// This will still be called even when the pop is canceled. A pop is canceled
+  /// when the relevant [Route.popEnabled] returns false, such as when
+  /// [popEnabled] is set to false on a [PopScope]. The `didPop` parameter
+  /// indicates whether or not the back navigation actually happened
+  /// successfully.
+  ///
+  /// See also:
+  ///
+  ///  * [Route.onPopped], which is similar.
+  @override
+  final OnPoppedCallback? onPopped;
+
+  /// {@template flutter.widgets.PopScope.popEnabled}
+  /// When false, blocks the current route from being popped.
+  ///
+  /// This includes the root route, where upon popping, the Flutter app would
+  /// exit.
+  ///
+  /// If multiple [PopScope] widgets appear in a route's widget subtree, then
+  /// each and every `popEnabled` must be `true` in order for the route to be
+  /// able to pop.
+  ///
+  /// [Android's predictive back](https://developer.android.com/guide/navigation/predictive-back-gesture)
+  /// feature will not animate when this boolean is false.
+  /// {@endtemplate}
+  @override
+  final bool popEnabled;
+
+  @override
+  State<PopScope> createState() => _PopScopeState();
+}
+
+class _PopScopeState extends State<PopScope> {
+  ModalRoute<dynamic>? _route;
+
+  @override
+  void didChangeDependencies() {
+    super.didChangeDependencies();
+    _route = ModalRoute.of(context);
+    _route?.registerPopInterface(widget);
+  }
+
+  @override
+  void didUpdateWidget(PopScope oldWidget) {
+    super.didUpdateWidget(oldWidget);
+    _route = ModalRoute.of(context);
+    _route?.unregisterPopInterface(oldWidget);
```

Comment:
This need to think about when people create bunch of const PopScope(), or the same widgets use elsewhere. in this case, if one unregister it would also remove registration of others.

Reply:
Oh good point hmm... Actually as-is, I haven't overridden `==`, so they will all be unique?  Maybe I should implement `==`?

If I did, then two PopScopes would only be equal if they had the same of every parameter, including `child` and `popEnabled`.  In that case, if they were in the same ModalRoute, it would be ok to register them twice, because `popEnabled` is the same.  If they were in different ModalRoutes, you would be registering them in different places, which is desired.

<details>

<summary>Some code to play around with this idea</summary>

```dart
main() {
  final Bar one = Bar(
    name: 'Justin',
    otherName: 'McCandless',
  );
  final Bar two = Bar(
    name: 'Justin',
    otherName: 'Bieber',
  );
  final Bar three = Bar(
    name: 'Justin',
    otherName: 'McCandless',
  );

  final Foo oneFoo = one;
  final Foo twoFoo = two;
  final Foo threeFoo = three;

  print('one == two ${one == two}, oneFoo == twoFoo ${oneFoo == twoFoo}');
  print('one == three ${one == three}, oneFoo == threeFoo ${oneFoo == threeFoo}');
}

sealed class Foo {
  const Foo({
    required this.name,
  });

  final String name;

  /*
  // This is never called because it's overridden in Bar.
  @override
  bool operator ==(Object other) {
    if (other.runtimeType != runtimeType) {
      return false;
    }
    return other is Foo
        && other.name == name;
  }
  */
}

class Bar implements Foo {
  const Bar({
    required this.name,
    required this.otherName,
  });

  final String name;
  final String otherName;

  /*
  @override
  bool operator ==(Object other) {
    if (other.runtimeType != runtimeType) {
      return false;
    }
    return other is Bar
        && other.name == name
        && other.otherName == otherName;
  }
  */
}
```

</details>

Reply:
for const object, they can be equal even if you don't have override `==`.
`const PopScope(popEnabled: true, Text('abc)) == const PopScope(popEnabled: true, Text('abc))`

registering is fine, but unregistering would become a problem. If both registered and one of them unregister itself, the other one will also be unregistered even if it doesn't call unregisterPopInterface

Reply:
Ah you're right, my problem was that I wasn't using `const` in my example code above (now edited).

So if two identical PopScopes are in the same ModalRoute, including having the same `child`, then unregistering one will unregister both.

Do you have any suggestion on how we should handle this?  I'm partially inclined to call this kind of thing user error and leave it as-is.  I'm not sure if other sort of register/unregister patterns handle this situation.  WidgetsBindingObserver is an example, though that's probably less likely to encounter a collision like this in practice:

https://github.com/flutter/flutter/blob/e1e2ebd176d6723a6063c47eea763f281723c885/packages/flutter/lib/src/widgets/binding.dart#L556

I could add a UniqueKey, or override hashCode, or something like that...  But it feels like I'd be making a weird change in order to accommodate a weird edge case.

I've added a test in pop_scope_test.dart that confirms the current behavior for now.

Reply:
i think we should avoid storing widget directly. maybe storing the state?

Reply:
That seems to have worked.  I made popEnabled into a ValueNotifier.  Let me know what you think.